conu/utils/core.py

```diff
@@ -33,68 +35,87 @@ def random_str():
-                os.makedirs(self.directory)
-        else:
-            self.directory = tempfile.mkdtemp()
-        if permissions:
-            self.set_permission(permissions)
-        if facl:
-            self.set_facl(facl)
-        if selinux_type:
-            self.set_selinux(selinux_type)
-        logger.debug("volume directory: %s" % self.directory)
-
-    def set_target(self, target):
-        self.target = target
-
-    def set_selinux(self, selinux_type):
-        run_cmd(["chcon", "-t", selinux_type, self.directory])
-
-    def set_force_selinux(self, value):
-        self.force_selinux = value
-
-    def set_permission(self, perms):
-        run_cmd(["chmod", perms, self.directory])
-
-    def set_facl(self, rules):
-        for rule in rules:
-            run_cmd(["setfacl", "-m", rule, self.directory])
+class Directory(object):
+    """
+    This class allows you to do advanced operations on filesystem directories, think of it
+    as mkdir on steroids
+    """
+    def __init__(self, path, mode=None, selinux_context=None, facl_rules=None):
+        """
+        :param path: str, path to the directory we will operate on
+        :param mode: int, octal representation of permission bits
+        :param selinux_context: str, set directory to this SELinux context
+        :param facl_rules: list of str, file ACLs to apply
+        """
+        # TODO: if path is None, we could do mkdtemp
+        self.path = path
+        self.mode = mode
+        self.context = selinux_context
+        self.facl_rules = facl_rules
+
+        # if set to True, then the directory is created and set up
+        self._initialized = False
+
+    def __enter__(self):
+        self.initialize()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.clean()
 
     def clean(self):
-        shutil.rmtree(self.directory)
+        if self._initialized:
+            logger.info("brace yourselves, removing %r", self.path)
+            shutil.rmtree(self.path)
+
+    def initialize(self):
+        """
+        create the directory if needed and configure it
+
+        :return: None
+        """
+        if not self._initialized:
+            logger.info("initializing %r", self)
+            if not os.path.exists(self.path):
+                if self.mode is not None:
+                    os.makedirs(self.path, mode=self.mode)
+                else:
+                    os.makedirs(self.path)
+            self._set_mode()
+            self._add_facl_rules()
+            self._set_selinux_context()
+            self._initialized = True
+            logger.info("initialized")
+            return
+        logger.info("%r was already initialized", self)
+
+    def _set_selinux_context(self):
+        # FIXME: do this using python API if possible
+        if self.context:
+            logger.debug("setting SELinux context of %s to %s", self.path, self.context)
+            run_cmd(["chcon", "-t", self.context, self.path])
```

Comment:
With -t switch only selinux type is changed. (As I understood selinux context = user:role:type:level) Name of function should be then set_selinux_type (or other should be added).